### DISCUSSION TITLE
How to create a player character

### DISCUSSION THREAD

**Discussion by ericmort** [nxrighthere/UnrealCLR#123] [Help] [answered]
First off, this is a great plugin! And I got the examples going with no problem, yay!
 
Now I am trying to create a basic UE project with a player character created in c# instead of c++. But I am unable to figure out how. Normally with c++ I can create a Blueprint class with my c++ class as parent and then assign that blueprint as DefaultPawnClass in my GameMode. But my C# class needs to inherit from my blueprint and UE doesn't know about any of my C# classes. There isnt't a GameMode version in C#. 

This seems so basic, so I feel like I must be doing something wrong here: how would you create a, say, FPS using CLR only? 

```csharp
class MyCharacter : Character
    {
        public MyCharacter() : base("myactor", Blueprint.Load("Blueprint'/Game/things/NewBlueprint.NewBlueprint'"))
        {
            Debug.AddOnScreenMessage(-1, 10.0f, Color.DeepPink, "Hello, Unreal Engine2!");


        }

        public void Print()
        {
            Debug.AddOnScreenMessage(-1, 10.0f, Color.DeepPink, "Hello, Unreal Engine3!");
        }
    }
```
```csharp
public static class Main
	{
		public static void OnWorldBegin()
		{
			//Debug.AddOnScreenMessage(-1, 10.0f, Color.DeepPink, "Hello, Unreal Engine1!");
			MyCharacter c = new MyCharacter();
			c.Print();

		}
```

Thanks!
Eric

**Comment by nxrighthere** [comment, marked as answer]
> But my C# class needs to inherit from my blueprint 

The path to the blueprint is incorrect, it should be just `Blueprint.Load("/Game/things/NewBlueprint")`.

> There isnt't a GameMode version in C#.

This API is not exposed yet, please, see the [roadmap](https://github.com/users/nxrighthere/projects/5?fullscreen=true).

**Reply by ericmort** [reply]
Thanks for replying, but I might not have been clear. I am trying to understand how to make the player character an instance of MyCharacter. For regular C++ I would set DefaultPawnClass etc. but since GameMode is not available in C#, how should I do it?

**Reply by nxrighthere** [reply]
You can retrieve a first player controller from the world using `World.GetFirstPlayerController()`, instantiate `MyCharacter` class, and call `playerController.Possess(myCharacter)`. This way you will get a controllable character.

**Reply by nxrighthere** [reply]
Player controllers can be swapped now using [`GameModeBase.SwapPlayerControllers()`](https://github.com/nxrighthere/UnrealCLR/blob/master/API/GameModeBase-SwapPlayerControllers(PlayerController_PlayerController).md). You can retrieve the [`GameModeBase`](https://github.com/nxrighthere/UnrealCLR/blob/master/API/GameModeBase.md) using [`World.GetGameMode()`](https://github.com/nxrighthere/UnrealCLR/blob/master/API/World-GetGameMode().md) method.

**Reply by ericmort** [reply]
Wow, thanks for the help and fast fix! I tried both those approaches but they give me an exception. 

```
public static class Main
	{ // Indicates the main entry point for automatic loading by the plugin
		public static void OnWorldBegin()
		{
			Debug.AddOnScreenMessage(-1, 10.0f, Color.DeepPink, "Hello, Unreal Engine1!");
			MyCharacter myCharacter = new MyCharacter();
			PlayerController pc = World.GetFirstPlayerController();
			Debug.AddOnScreenMessage(-1, 10.0f, Color.DeepPink, "Got Player controller");
			pc.Possess(myCharacter);

			GameModeBase gm = World.GetGameMode();
			Debug.AddOnScreenMessage(-1, 10.0f, Color.DeepPink, "Got game mode");
			gm.SwapPlayerControllers(World.GetFirstPlayerController(),pc);
			Debug.AddOnScreenMessage(-1, 10.0f, Color.DeepPink, "Swapped player controllers");

		}

```

I get:

```
LogUnrealCLR: Error: UnrealCLR::Module::Exception: System.NullReferenceException: Object reference not set to an instance of an object.
   at Game.Main.OnWorldBegin()
   at UnrealEngine.Runtime.Core.ManagedCommand(Command command) in C:\Users\ericm\Downloads\UnrealCLR-master\UnrealCLR-master\Source\Managed\Runtime\Runtime.cs:line 150
```

It turns out both World.GetFirstPlayerController() and World.GetGameMode() return null.

**Reply by ericmort** [reply]
Don't worry about it, that last part was my mistake! I got it working now, at least I got the code above to run without exceptions. thanks a lot!
